engine/render/src/render/program_utils.cpp

```diff
@@ -236,23 +248,38 @@ namespace dmRender
             if (uniform_desc.m_Type == dmGraphics::TYPE_FLOAT_VEC4 || uniform_desc.m_Type == dmGraphics::TYPE_FLOAT_MAT4)
             {
                 HConstant render_constant = dmRender::NewConstant(uniform_desc.m_NameHash);
+
                 dmRender::SetConstantLocation(render_constant, uniform_desc.m_Location);
 
-                if (uniform_desc.m_Type == dmGraphics::TYPE_FLOAT_MAT4)
+                // We are be about to add a duplicate. Make sure to reuse the data
+                HConstant prev_constant = FindConstant(constants, uniform_desc.m_NameHash);
+                if (prev_constant != 0)
                 {
-                    num_values *= 4;
-                    dmRender::SetConstantType(render_constant, dmRenderDDF::MaterialDesc::CONSTANT_TYPE_USER_MATRIX4);
-                }
+                    uint32_t prev_num_values;
+                    dmVMath::Vector4* prev_values = GetConstantValues(prev_constant, &prev_num_values);
 
-                // Set correct size of the constant (Until the shader builder provides all the default values)
-                if (num_values > default_values_capacity)
+                    dmRender::SetConstantValuesRef(render_constant, prev_values, prev_num_values);
+                    dmRender::SetConstantType(render_constant, GetConstantType(prev_constant));
```

Comment:
I do believe we need to check mor thoroughly in editor/bob for mismatching types etc.
For now, we use the type of the first uniform (as we are using an existing pointer to the data)

Reply:
Could you please create a task for bob?